### PR TITLE
fix(typing) Improve types for env()

### PR DESCRIPTION
**Pull request by markstory**
This wrapper only accepts our homebrew type coercion instances. If one were to use `bool` the parsed value will be wrong.

### CODE REVIEW

File: src/sentry/conf/server.py [hunk 1]
```diff
@@ -32,20 +41,24 @@ def gettext_noop(s: str) -> str:
 socket.setdefaulttimeout(5)
 
 
+T = TypeVar("T")
```
Comment (corps) [inline src/sentry/conf/server.py:44]: minor nit: using `_T` here prevents autocomplete from trying to import these across modules.

**Review by corps — APPROVED** [review]

File: src/sentry/utils/types.py [hunk 2]
```diff
@@ -181,7 +181,7 @@ def convert(self, value):
 }
 
 
-def type_from_value(value: typing.Any) -> typing.Callable[[typing.Any], typing.Any]:
+def type_from_value(value: typing.Any) -> Type:
```
Comment (bukzor-sentryio) [inline src/sentry/utils/types.py:184]: the typing system could understand this completely if you add some overrides:

```suggestion
@typing.overload
def type_from_value(value: bytes) -> StringType:
    ...


@typing.overload
def type_from_value(value: str) -> StringType:
    ...


@typing.overload
def type_from_value(value: list[T]) -> SequenceType[T]:
    ...


def type_from_value(value: typing.Any) -> Type:
    """Fetch Type based on a primitive value"""
    return _type_mapping[type(value)]


reveal_type(type_from_value([3]))  # revealed type is demo.SequenceType[int]
```

Without the overrides it gets `demo.Type[Any]`, instead.

File: src/sentry/conf/server.py (hunk 1 above)
Comment (markstory) [reply src/sentry/conf/server.py:44]: I can update that.

File: src/sentry/conf/server.py [hunk 3]
```diff
@@ -32,20 +41,24 @@ def gettext_noop(s: str) -> str:
...
+def env(key: str, default: T, type: Optional[TypeParser] = None) -> T:
     ...
 
 
 def env(
     key: str,
     default: str | T = "",
-    type: Optional[Callable[[Any], T]] = None,
+    type: Optional[TypeParser] = None,
 ) -> T:
```
Comment (bukzor-sentryio) [inline src/sentry/conf/server.py]: This new function signature only mentions the typevar `TypeParser` once, which makes it unable to constrain any calls to this function. If the intent is to just constrain T to `sentry.utils.types.Type` i'd do this:

```
from sentry.utils.types import Type
T_Type = TypeVar("T_Type", bound=Type)

def env(
    key: str,
    default: T_Type | str = "",
    type: T_Type | None = None,
) -> T_Type:

```

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/sentry/utils/types.py (hunk 2 above)
Comment (markstory) [reply src/sentry/utils/types.py]: Thanks for the suggestion on adding overloads for type_from_value(). I've added those now, with the exception of `type_from_value(value: bool) -> BoolType` as mypy thinks it overlaps with the `int` overload 🙃

**Review by vbro — APPROVED** [review]

File: src/sentry/utils/types.py [hunk 4]
```diff
@@ -181,7 +181,41 @@ def convert(self, value):
 }
 
 
-def type_from_value(value: typing.Any) -> typing.Callable[[typing.Any], typing.Any]:
+# @typing.overload
+# def type_from_value(value: bool) -> BoolType:
```
Comment (bukzor-sentryio) [inline src/sentry/utils/types.py:185]: what's wrong with bool?

**Review by ghost — APPROVED** [review]

File: src/sentry/utils/types.py (hunk 2 above)
Comment (bukzor-sentryio) [reply src/sentry/utils/types.py]: leave a comment? it may be something we can improve later.

File: src/sentry/conf/server.py [hunk 5]
```diff
@@ -33,27 +31,31 @@ def gettext_noop(s: str) -> str:
...
 
 
 def env(
     key: str,
-    default: str | T = "",
-    type: Optional[Callable[[Any], T]] = None,
-) -> T:
+    default: str | _EnvTypes = "",
+    type: Optional[Type] = None,
+) -> _EnvTypes:
```
Comment (bukzor-sentryio) [inline src/sentry/conf/server.py:51]: This no longer constrains env to have a matching default and return value, which isn't what you intended, I think. What was wrong with my suggestion from before?

File: src/sentry/utils/types.py (hunk 4 above)
Comment (markstory) [reply src/sentry/utils/types.py:185]: If this was left uncommented mypy would complain about overloads conflicting. I don't understand why though as int and bool are separate types.

Comment (mrduncan) [reply src/sentry/utils/types.py:185]: Related to python's [weird bool inheritance](https://peps.python.org/pep-0285/) maybe?

```python
>>> isinstance(True, int)
True
```

Comment (bukzor-sentryio) [reply src/sentry/utils/types.py:185]: This works well:
```python
import typing


class A: pass
class B: pass

@typing.overload
# type: ignore[misc]  # bool overlaps with int
def f(x: bool) -> B: ...
@typing.overload
def f(x: int) -> A: ...

def f(x):
    if isinstance(x, bool): return A()
    elif isinstance(x, str): return B()
    else: raise TypeError(type(x))

reveal_type(f(1))   # Revealed type is "demo.A"
reveal_type(f(True))  # Revealed type is "demo.B"
```

As far as I can see, this error is plain old wrong, and things start working correctly with it disabled.

**Review by ghost — APPROVED** [review]

**Pull request merged**